src/backend/distributed/commands/multi_copy.c

```diff
@@ -623,196 +642,6 @@ CopyToExistingShards(CopyStmt *copyStatement, QueryCompletionCompat *completionT
-		bool nextRowFound = NextCopyFromCompat(copyState, executorExpressionContext,
-											   columnValues, columnNulls);
-
-		if (!nextRowFound)
-		{
-			/* switch to regular memory context and stop showing line number in errors */
-			MemoryContextSwitchTo(oldContext);
-			error_context_stack = errorCallback.previous;
-			break;
-		}
-
-		CHECK_FOR_INTERRUPTS();
-
-		/* switch to regular memory context and stop showing line number in errors */
-		MemoryContextSwitchTo(oldContext);
-		error_context_stack = errorCallback.previous;
-
-		/*
-		 * If copied data size is zero, this means either this is the first
-		 * line in the copy or we just filled the previous shard up to its
-		 * capacity. Either way, we need to create a new shard and
-		 * start copying new rows into it.
-		 */
-		if (copiedDataSizeInBytes == 0)
-		{
-			/* create shard and open connections to shard placements */
-			currentShardId = StartCopyToNewShard(shardConnections, copyStatement,
-												 copyOutState->binary);
-
-			/* send copy binary headers to shard placements */
-			if (copyOutState->binary)
-			{
-				SendCopyBinaryHeaders(copyOutState, currentShardId,
-									  shardConnections->connectionList);
-			}
-		}
-
-		/* replicate row to shard placements */
-		resetStringInfo(copyOutState->fe_msgbuf);
-		AppendCopyRowData(columnValues, columnNulls, tupleDescriptor,
-						  copyOutState, columnOutputFunctions, NULL);
-		SendCopyDataToAll(copyOutState->fe_msgbuf, currentShardId,
-						  shardConnections->connectionList);
-
-		uint64 messageBufferSize = copyOutState->fe_msgbuf->len;
-		copiedDataSizeInBytes = copiedDataSizeInBytes + messageBufferSize;
-
-		/*
-		 * If we filled up this shard to its capacity, send copy binary footers
-		 * to shard placements, and update shard statistics.
-		 */
-		if (copiedDataSizeInBytes > shardMaxSizeInBytes)
-		{
-			Assert(currentShardId != INVALID_SHARD_ID);
-
-			if (copyOutState->binary)
-			{
-				SendCopyBinaryFooters(copyOutState, currentShardId,
-									  shardConnections->connectionList);
-			}
-
-			EndRemoteCopy(currentShardId, shardConnections->connectionList);
-			UpdateShardStatistics(shardConnections->shardId);
-
-			copiedDataSizeInBytes = 0;
-			currentShardId = INVALID_SHARD_ID;
-		}
-
-		processedRowCount += 1;
-
-#if PG_VERSION_NUM >= PG_VERSION_14
-		pgstat_progress_update_param(PROGRESS_COPY_TUPLES_PROCESSED, processedRowCount);
-#endif
-	}
-
-	/*
-	 * For the last shard, send copy binary footers to shard placements,
-	 * and update shard statistics. If no row is send, there is no shard
-	 * to finalize the copy command.
-	 */
-	if (copiedDataSizeInBytes > 0)
-	{
-		Assert(currentShardId != INVALID_SHARD_ID);
-
-		if (copyOutState->binary)
-		{
-			SendCopyBinaryFooters(copyOutState, currentShardId,
-								  shardConnections->connectionList);
-		}
-		EndRemoteCopy(currentShardId, shardConnections->connectionList);
-		UpdateShardStatistics(shardConnections->shardId);
```

Comment:
That's intentional. I don't want COPY to change (and have to sync) metadata, since I'd also expect shards without shardminvalue/shardmaxvalue to take arbitrary inserts.